### ISSUE TITLE
Exitcode problem

### ISSUE THREAD

**Issue by Selur** [rigaya/NVEnc#189]
Starting with 4.60 I always get a non-zero exit code (exitCode -1073741819) when calling NVEncC (using QProcess from the Qt Framework). (Also tried current 4.61, same issue.)

Everything works fine with NVEncC 4.59 and before.
-> would be nice if this could be fixed.

**Comment by rigaya** [comment]
Would you let me know the exact command line and the log?

I've been testing with various options, but I get exit code 0, when there is no error.

Actually, -1073741819 is 0x‭C0000005‬ in hex, which means segmentation fault, the problem might not only the error code.

**Comment by Selur** [comment]
Encoding call:
```
"I:\Hybrid\64bit\NVEncC64.exe" --avhw  -i "F:\TestClips&Co\files\5000frames.mp4" --fps 25.000 --codec h265 --profile main10 --level auto --tier high --sar 1:1 --lookahead 32 --output-depth 10 --vbrhq 0 --vbr-quality 18.00 --max-bitrate 240000 --aq --aq-strength 3 --gop-len 0 --ref 3 --multiref-l0 3 --multiref-l1 3 --nonrefp --bframes 0 --no-b-adapt --mv-precision Q-pel --preset quality --colorprim undef --transfer undef --colormatrix bt470bg --vpp-resize spline36 --output-res 640x480 --vpp-gauss disabled --vpp-unsharp radius=2,weight=0.5,threshold=10 --cuda-schedule sync --output "E:\Temp\5000frames_08_14_34_5110_01.265"
```
Uploaded the source to: https://drive.google.com/open?id=1dqd2wwsfXG3fxum2VWyUXClutUkchEDQ

Outputs:
```
Multiple Refs unsupported.
NVEncC (x64) 4.61 (r1307) by rigaya, Jan 16 2020 14:49:35 (VC 1916/Win/avx2)
OS Version     Windows 10 x64 (18363)
CPU            AMD Ryzen 7 1800X Eight-Core Processor (8C/16T)
GPU            #0: GeForce GTX 1070 Ti (2432 cores, 1683 MHz)[PCIe3x16][441.87]
NVENC / CUDA   NVENC API 9.1, CUDA 10.2, schedule mode: sync
Input Buffers  CUDA, 41 frames
Input Info     avcuvid: H.265/HEVC, 640x480, 25/1 fps
Vpp Filters    cspconv(p010 -> yv12(16bit))
               unsharp: radius 2, weight 0.5, threshold 10.0
               cspconv(yv12(16bit) -> p010)
Output Info    H.265/HEVC main10 @ Level auto
               640x480p 1:1 25.000fps (25/1fps)
Encoder Preset quality
Rate Control   VBRHQ
Bitrate        0 kbps (Max: 240000 kbps)
Target Quality 18.00
Initial QP     I:20  P:23  B:25
VBV buf size   auto
Lookahead      on, 32 frames, Adaptive I Insert
GOP length     250 frames
B frames       0 frames [ref mode: disabled]
Ref frames     3 frames
AQ             on
CU max / min   auto / auto
Others         mv:Q-pel nonrefp 
[15.4%] 656 frames: 1115.65 fps, 37 kb/s, remain 0:00:03, GPU 5%, VE 0%, est out size 0.8MB
[34.0%] 1642 frames: 1182.15 fps, 36 kb/s, remain 0:00:03, GPU 20%, VE 49%, VD 23%, est out size 0.8MB
[52.6%] 2632 frames: 1202.38 fps, 36 kb/s, remain 0:00:02, GPU 21%, VE 50%, VD 24%, est out size 0.9MB
[71.2%] 3622 frames: 1211.78 fps, 35 kb/s, remain 0:00:01, GPU 21%, VE 50%, VD 24%, est out size 0.9MB
[89.8%] 4609 frames: 1216.42 fps, 35 kb/s, remain 0:00:00, GPU 20%, VE 49%, VD 24%, est out size 0.9MB
```
at the end I get:
```
process finished with exitcode: -1073741819 and exitStatue: 1
```
using the same code and just replacing the NVEncC64 binary (and the support files) I get:
```
Multiple Refs unsupported.
NVEncC (x64) 4.59 (r1270) by rigaya, Dec 24 2019 13:48:23 (VC 1916/Win/avx2)
OS Version     Windows 10 x64 (18363)
CPU            AMD Ryzen 7 1800X Eight-Core Processor (8C/16T)
GPU            #0: GeForce GTX 1070 Ti (2432 cores, 1683 MHz)[PCIe3x16][441.87]
NVENC / CUDA   NVENC API 9.1, CUDA 10.2, schedule mode: sync
Input Buffers  CUDA, 41 frames
Input Info     avcuvid: H.265/HEVC, 640x480, 25/1 fps
Vpp Filters    cspconv(p010 -> yv12(16bit))
               unsharp: radius 2, weight 0.5, threshold 10.0
               cspconv(yv12(16bit) -> p010)
Output Info    H.265/HEVC main10 @ Level auto
               640x480p 1:1 25.000fps (25/1fps)

Encoder Preset quality
Rate Control   VBRHQ
Bitrate        0 kbps (Max: 240000 kbps)
Target Quality 18.00
Initial QP     I:20  P:23  B:25
VBV buf size   auto
Lookahead      on, 32 frames, Adaptive I Insert
GOP length     250 frames
B frames       0 frames [ref mode: disabled]
Ref frames     3 frames
AQ             on
CU max / min   auto / auto
Others         mv:Q-pel nonrefp 
[15.7%] 659 frames: 1091.21 fps, 37 kb/s, remain 0:00:03, GPU 6%, VE 0%, est out size 0.7MB
[34.3%] 1651 frames: 1176.05 fps, 36 kb/s, remain 0:00:03, GPU 20%, VE 50%, est out size 0.8MB
[52.9%] 2641 frames: 1198.82 fps, 36 kb/s, remain 0:00:02, GPU 20%, VE 50%, est out size 0.8MB
[71.2%] 3629 frames: 1208.46 fps, 35 kb/s, remain 0:00:01, GPU 20%, VE 50%, est out size 0.9MB
[89.8%] 4619 frames: 1214.57 fps, 35 kb/s, remain 0:00:00, GPU 21%, VE 49%, est out size 0.9MB
encoded 5000 frames, 1217.73 fps, 35.19 kbps, 0.84 MB
encode time 0:00:04, CPU: 4.0, GPU: 17.4, VE: 39.8, GPUClock: 1820MHz, VEClock: 1635MHz
frame type IDR   20
frame type I     20,  total size  0.01 MB
frame type P   4980,  total size  0.83 MB
```

this happens for me with any source, 4.59 works fine, 4.60+ gives me the "exitcode: -1073741819"

**Comment by rigaya** [comment]
Thank you for the samples, but although I have used a similar system, I was unsuccessful in reproducing the issue.

cmd with the same options: ERRORLEVEL is 0 after run.

![cmd 2020_01_19 18_13_46](https://user-images.githubusercontent.com/815947/72678676-8ddcba00-3aeb-11ea-8b90-66498f168427.png)

I've also tested using Hybrid but nothing weird happened.
```
NVEnc processing started...
starting 18_34_08_1210_01_video@18:34:08.966 - F:\temp\test.mp4
"Y:\QSVTest\x64\NVEncC64.exe" --avhw  -i "F:\temp\5000frames.mp4" --fps 25.000 --codec h265 --profile main10 --level auto --tier high --sar 1:1 --lookahead 32 --output-depth 10 --vbrhq 0 --vbr-quality 18.00 --max-bitrate 240000 --aq --gop-len 0 --ref 3 --multiref-l0 3 --multiref-l1 3 --bframes 0 --no-b-adapt --mv-precision Q-pel --preset default --colormatrix bt470bg --vpp-resize default --output-res 640x480 --vpp-gauss disabled --vpp-unsharp radius=2,weight=0.5,threshold=10 --cuda-schedule sync --output "C:\Users\rigaya\AppData\Local\Temp\test_18_34_08_1210_01.265"
encoded 5000 frames, 1316.14 fps, 34.03 kbps, 0.81 MB
encode time 0:00:03, CPU: 3.0, GPU: 31.2, VE: 83.6, VD: 21.4, GPUClock: 1865MHz, VEClock: 1673MHz
frame type IDR   20
frame type I     20,  total size  0.01 MB
frame type P   4980,  total size  0.80 MB
finished after 00:00:07.477
finished...

created C:\Users\rigaya\AppData\Local\Temp\test_18_34_08_1210_01.265 (0.81144 MB)
starting 18_34_08_1210_02_muxing@18:34:16.775 - F:\temp\test.mp4
"C:\PROGRA~1\Hybrid\64bit\MP4Box.exe" -add "C:\Users\rigaya\AppData\Local\Temp\TEST_1~1.265"#video:fps=25:xps_inband:name="" -brand hev1 -itags tool="Hybrid 2020.01.11.1" -tmp "C:\Users\rigaya\AppData\Local\Temp" -new "F:\temp\18_34_08_1210__02.mp4"
finished after 00:00:00.150
finished...

created F:\temp\18_34_08_1210__02.mp4 (0.834485 MB)
```



Would you please have a run using the binaries below adding "--log-level debug"?
[NVEncC64_4.61_test1.zip](https://github.com/rigaya/NVEnc/files/4082504/NVEncC64_4.61_test1.zip)

You should get something like below at the end of the log, if NVEncC64 crushes somewhere, that might be a hint of the problem.
```
avcuvid: 5000 frames, End of file
Flushing Decoder
Flushed Decoder
Flushed Encoder
bitstream: Closing...
bitstream: Closed file pointer.
bitstream: Closed.
avcuvid: Closing...
avcuvid: Closed Stream Packet Buffer.
avcuvid: Closed caption handler.
avcuvid: Closing file pointer...
avcuvid: Closing pb->buffer...
avcuvid: Closed pb->buffer.
avcuvid: Closing avio context...
avcuvid: Closed avio context.
avcuvid: Closed file pointer.
avcuvid: Closing avformat context...
avcuvid: Closed avformat context.
avcuvid: Free formatOptions...
avcuvid: Freed formatOptions.
avcuvid: Closed format.
avcuvid: Close parser...
avcuvid: Closed parser.
avcuvid: Close codecCtx for parser...
avcuvid: Closed codecCtx for parser.
avcuvid: Free extra data...
avcuvid: Freed extra data.
avcuvid: Closed video.
avcuvid: Cleared frame pos list.
avcuvid: Closed.

encoded 5000 frames, 585.89 fps, 25.73 kbps, 0.61 MB
encode time 0:00:08, CPU: 0.9%, GPU: 20.0%, VE: 89.7%, VD: 7.3%, GPUClock: 1833MHz, VEClock: 1700MHz
frame type IDR   20
frame type I     20,  total size  0.01 MB
frame type P   4980,  total size  0.60 MB
avcuvid: Closing...
avcuvid: Closed Stream Packet Buffer.
avcuvid: Closed caption handler.
avcuvid: Closed format.
avcuvid: Closed video.
avcuvid: Cleared frame pos list.
avcuvid: Closed.
avcuvid: Closing...
avcuvid: Close...
nvEncDestroyEncoder: success.
Closing perf monitor...
```

**Comment by Selur** [comment]
Uploaded the output to:
https://drive.google.com/open?id=1W9VxoQEALkXCE5yS7K6EmS7gwX4zlUVW
I don't get the stuff at the end of the log at all. :/
Always at the end it seems like NVEncC 'freezes' and then I get the error, I first thought that I might have changed something on the code that is calling NVEncC bit I didn't and like I wrote 4.59 works fine. :/

**Comment by Selur** [comment]
I do get all the output when I call NVEncC inside a windows command prompt,..
```
avcuvid: 5000 frames, End of file
Flushing Decoder
Flushed Decoder
Flushed Encoder
bitstream: Closing...
bitstream: Closed file pointer.
bitstream: Closed.
avcuvid: Closing...
avcuvid: Closed Stream Packet Buffer.
avcuvid: Closed caption handler.
avcuvid: Closing file pointer...
avcuvid: Closing pb->buffer...
avcuvid: Closed pb->buffer.
avcuvid: Closing avio context...
avcuvid: Closed avio context.
avcuvid: Closed file pointer.
avcuvid: Closing avformat context...
avcuvid: Closed avformat context.
avcuvid: Free formatOptions...
avcuvid: Freed formatOptions.
avcuvid: Closed format.
avcuvid: Close parser...
avcuvid: Closed parser.
avcuvid: Close codecCtx for parser...
avcuvid: Closed codecCtx for parser.
avcuvid: Free extra data...
avcuvid: Freed extra data.
avcuvid: Closed video.
avcuvid: Cleared frame pos list.
avcuvid: Closed.

encoded 5000 frames, 1055.30 fps, 35.19 kbps, 0.84 MB
encode time 0:00:04, CPU: 4.3%, GPU: 18.0%, VE: 27.0%, VD: 13.2%, GPUClock: 1792MHz, VEClock: 1611MHz
frame type IDR   20
frame type I     20,  total size  0.01 MB
frame type P   4980,  total size  0.83 MB
avcuvid: Closing...
avcuvid: Closed Stream Packet Buffer.
avcuvid: Closed caption handler.
avcuvid: Closed format.
avcuvid: Closed video.
avcuvid: Cleared frame pos list.
avcuvid: Closed.
avcuvid: Closing...
avcuvid: Close...
nvEncDestroyEncoder: success.
Closing perf monitor...
```
but I don't get it anymore when using Qt to call NVEncC

**Comment by Selur** [comment]
When adding: "--log "e:\Output.txt" and "--log-level debug"  to the call, the Output.txt is written and properly ends with:
```
avcuvid: 5000 frames, End of file
Flushing Decoder
Flushed Decoder
Flushed Encoder
bitstream: Closing...
bitstream: Closed file pointer.
bitstream: Closed.
avcuvid: Closing...
avcuvid: Closed Stream Packet Buffer.
avcuvid: Closed caption handler.
avcuvid: Closing file pointer...
avcuvid: Closing pb->buffer...
avcuvid: Closed pb->buffer.
avcuvid: Closing avio context...
avcuvid: Closed avio context.
avcuvid: Closed file pointer.
avcuvid: Closing avformat context...
avcuvid: Closed avformat context.
avcuvid: Free formatOptions...
avcuvid: Freed formatOptions.
avcuvid: Closed format.
avcuvid: Close parser...
avcuvid: Closed parser.
avcuvid: Close codecCtx for parser...
avcuvid: Closed codecCtx for parser.
avcuvid: Free extra data...
avcuvid: Freed extra data.
avcuvid: Closed video.
avcuvid: Cleared frame pos list.
avcuvid: Closed.
                                                                               
encoded 5000 frames, 1226.69 fps, 35.19 kbps, 0.84 MB
encode time 0:00:04, CPU: 4.4%, GPU: 16.5%, VE: 37.5%, VD: 18.0%, GPUClock: 1816MHz, VEClock: 1632MHz
frame type IDR   20
frame type I     20,  total size  0.01 MB
frame type P   4980,  total size  0.83 MB
avcuvid: Closing...
avcuvid: Closed Stream Packet Buffer.
avcuvid: Closed caption handler.
avcuvid: Closed format.
avcuvid: Closed video.
avcuvid: Cleared frame pos list.
avcuvid: Closed.
avcuvid: Closing...
avcuvid: Close...
nvEncDestroyEncoder: success.
```
but I still get the  "exitcode: -1073741819"

**Comment by rigaya** [comment]
> I do get all the output when I call NVEncC inside a windows command prompt,..
> but I don't get it anymore when using Qt to call NVEncC

Are there differences between them? I'm not familiar to Qt...

Thank you for adding the --log option, the output log will exactly show where NVEncC crushed. After "nvEncDestroyEncoder: success." should come "Closing perf monitor...", so it seems like NVEncC crushed between them.

I've created another test build to check, would you please have a run adding "--log "e:\Output.txt" and "--log-level debug"? Result should be like below.

```
encoded 5000 frames, 586.44 fps, 25.73 kbps, 0.61 MB
encode time 0:00:08, CPU: 0.9%, GPU: 20.3%, VE: 90.0%, VD: 7.3%, GPUClock: 1827MHz, VEClock: 1698MHz
frame type IDR   20
frame type I     20,  total size  0.01 MB
frame type P   4980,  total size  0.60 MB
avcuvid: Closing...
avcuvid: Closed Stream Packet Buffer.
avcuvid: Closed caption handler.
avcuvid: Closed format.
avcuvid: Closed video.
avcuvid: Cleared frame pos list.
avcuvid: Closed.
avcuvid: Closing...
avcuvid: Close...
nvEncDestroyEncoder: success.
cuvid: Closing decoder...
cuvid: cuvidDestroyDecoder: Fin.
cuvid: cuvidDestroyVideoParser: Fin.
cuvid: Closed decoder.
cuvidCtxLockDestroy...
cuvidCtxLockDestroy: Fin.
Closing EncodeStatus...
Closed EncodeStatus.
cuCtxDestroy...
cuCtxDestroy: Fin.
Closing perf monitor...
perf monitor: Closing thread...
perf monitor: Closed thread.
perf monitor: Closing perf counter...
perf monitor: Closed perf counter.
Closing logger...

Y:\QSVTest>echo %ERRORLEVEL%
0
```

[NVEncC64_4.61_test2.zip](https://github.com/rigaya/NVEnc/files/4082767/NVEncC64_4.61_test2.zip)

**Comment by Selur** [comment]
I get:
```
avcuvid: 5000 frames, End of file
Flushing Decoder
Flushed Decoder
Flushed Encoder
bitstream: Closing...
bitstream: Closed file pointer.
bitstream: Closed.
avcuvid: Closing...
avcuvid: Closed Stream Packet Buffer.
avcuvid: Closed caption handler.
avcuvid: Closing file pointer...
avcuvid: Closing pb->buffer...
avcuvid: Closed pb->buffer.
avcuvid: Closing avio context...
avcuvid: Closed avio context.
avcuvid: Closed file pointer.
avcuvid: Closing avformat context...
avcuvid: Closed avformat context.
avcuvid: Free formatOptions...
avcuvid: Freed formatOptions.
avcuvid: Closed format.
avcuvid: Close parser...
avcuvid: Closed parser.
avcuvid: Close codecCtx for parser...
avcuvid: Closed codecCtx for parser.
avcuvid: Free extra data...
avcuvid: Freed extra data.
avcuvid: Closed video.
avcuvid: Cleared frame pos list.
avcuvid: Closed.
                                                                               
encoded 5000 frames, 1225.49 fps, 35.19 kbps, 0.84 MB
encode time 0:00:04, CPU: 4.3%, GPU: 16.0%, VE: 37.5%, VD: 18.0%, GPUClock: 1816MHz, VEClock: 1632MHz
frame type IDR   20
frame type I     20,  total size  0.01 MB
frame type P   4980,  total size  0.83 MB
avcuvid: Closing...
avcuvid: Closed Stream Packet Buffer.
avcuvid: Closed caption handler.
avcuvid: Closed format.
avcuvid: Closed video.
avcuvid: Cleared frame pos list.
avcuvid: Closed.
avcuvid: Closing...
avcuvid: Close...
nvEncDestroyEncoder: success.
cuvid: Closing decoder...
cuvid: cuvidDestroyDecoder: Fin.
cuvid: cuvidDestroyVideoParser: Fin.
cuvid: Closed decoder.
cuvidCtxLockDestroy...
cuvidCtxLockDestroy: Fin.
Closing EncodeStatus...
Closed EncodeStatus.
cuCtxDestroy...
cuCtxDestroy: Fin.
Closing perf monitor...
perf monitor: Closing thread...
perf monitor: Closed thread.
perf monitor: Closing perf counter...
```
the 'Closing logger...' part does not appear here. 
to check I also used the call in the Windows command line and there too 'Closing logger...' does not appear.

**Comment by rigaya** [comment]
Thanks a lot, I think I've got where it crush.

I've made a build which probably will fix the crush, please have a try.
[NVEncC64_4.61_test3.zip](https://github.com/rigaya/NVEnc/files/4082808/NVEncC64_4.61_test3.zip)

**Comment by Selur** [comment]
I can confirm that build fixes the crash. :)
and the debug output shows:
```
encoded 5000 frames, 1203.95 fps, 35.19 kbps, 0.84 MB
encode time 0:00:04, CPU: 4.6%, GPU: 17.6%, VE: 38.6%, VD: 18.2%, GPUClock: 1829MHz, VEClock: 1644MHz
frame type IDR   20
frame type I     20,  total size  0.01 MB
frame type P   4980,  total size  0.83 MB
avcuvid: Closing...
avcuvid: Closed Stream Packet Buffer.
avcuvid: Closed caption handler.
avcuvid: Closed format.
avcuvid: Closed video.
avcuvid: Cleared frame pos list.
avcuvid: Closed.
avcuvid: Closing...
avcuvid: Close...
nvEncDestroyEncoder: success.
cuvid: Closing decoder...
cuvid: cuvidDestroyDecoder: Fin.
cuvid: cuvidDestroyVideoParser: Fin.
cuvid: Closed decoder.
cuvidCtxLockDestroy...
cuvidCtxLockDestroy: Fin.
Closing EncodeStatus...
Closed EncodeStatus.
cuCtxDestroy...
cuCtxDestroy: Fin.
Closing perf monitor...
perf monitor: Closing thread...
perf monitor: Closed thread.
perf monitor: Closing perf counter...
perf monitor: Closed perf counter.
Closing logger...
```
Thanks!

**Comment by rigaya** [comment]
Thank you for your cooperation! I’m glad to hear that it works. I’ll include this fix in the next release.

**Comment by rigaya** [comment]
The fix has been applied to [NVEnc 4.62](https://github.com/rigaya/NVEnc/releases/tag/4.62).
